pkg/archive/parser.go

```diff
@@ -165,3 +167,102 @@ func (p *Parser) getResourceItemsForScope(dir, archiveRootDir string) ([]string,
 
 	return items, nil
 }
+
+// CheckAndReadDir is a wrapper around fs.DirExists and fs.ReadDir that does checks
+// and returns errors if directory cannot be read.
+func (p *Parser) CheckAndReadDir(dir string) ([]os.FileInfo, error) {
+	exists, err := p.fs.DirExists(dir)
+	if err != nil {
+		return []os.FileInfo{}, errors.Wrapf(err, "finding %q", dir)
+	}
+	if !exists {
+		return []os.FileInfo{}, errors.Errorf("%q not found", dir)
+	}
+
+	contents, err := p.fs.ReadDir(dir)
+	if err != nil {
+		return []os.FileInfo{}, errors.Wrapf(err, "reading contents of %q", dir)
+	}
+
+	return contents, nil
+}
+
+// ParseGroupVersions extracts the versions for each API Group from the backup
+// directory names and stores them in a metav1 APIGroup object.
+func (p *Parser) ParseGroupVersions(dir string) (map[string]metav1.APIGroup, error) {
+	resourcesDir := filepath.Join(dir, velerov1api.ResourcesDir)
+
+	// Get the subdirectories inside the "resources" directory. The subdirectories
+	// will have resource.group names like "horizontalpodautoscalers.autoscaling".
+	rgDirs, err := p.CheckAndReadDir(resourcesDir)
+	if err != nil {
+		return nil, err
+	}
+
+	resourceAGs := make(map[string]metav1.APIGroup)
+
+	// Loop through the resource.group directory names.
+	for _, rgd := range rgDirs {
+		group := metav1.APIGroup{
+			Name: extractGroupName(rgd.Name()),
+		}
+
+		rgdPath := filepath.Join(resourcesDir, rgd.Name())
+
+		// Inside each of the resource.group directories are directories whose
+		// names are API Group versions like "v1" or "v1-preferredversion"
+		gvDirs, err := p.CheckAndReadDir(rgdPath)
+		if err != nil {
+			return nil, err
+		}
+
+		var supportedVersions []metav1.GroupVersionForDiscovery
+
+		for _, gvd := range gvDirs {
+			gvdName := gvd.Name()
+
+			// Don't save the namespaces or clusters directories in list of
+			// supported API Group Versions.
+			if gvdName == "namespaces" || gvdName == "cluster" {
+				continue
+			}
+
+			version := metav1.GroupVersionForDiscovery{
+				GroupVersion: strings.TrimPrefix(group.Name+"/"+gvdName, "/"),
+				Version:      gvdName,
+			}
+
+			if strings.Contains(gvdName, "-preferredversion") {
```

Comment:
We should use the constant PreferredVersionDir

Reply:
Will do.